File: backend/src/Core/Features/Tenants/Commands/ToggleTenantStatus.cs
namespace Core.Features.Tenants.Commands;

public static class ToggleTenantStatus
{
    public class Validator : AbstractValidator<Command>
    {
        public Validator(ITenantRepository tenantRepository)
        {
            RuleFor(x => x.Code)
                .Cascade(CascadeMode.Stop)
                .ExistsTenant(tenantRepository);
        }
    }

    public class Command(int code, bool isEnabled) : IRequest<TenantStatusDto>
    {
        public int Code { get; set; } = code;

        public bool IsEnabled { get; } = isEnabled;
    }

    public class Handler(ITenantRepository tenantRepository) : IRequestHandler<Command, TenantStatusDto>
    {
        public async Task<TenantStatusDto> Handle(Command request, CancellationToken cancellationToken)
        {
            var updateDone = await tenantRepository.UpdateAsync(request.Code, new UpdateTenantProperties(isEnabled: request.IsEnabled), cancellationToken);

            return new TenantStatusDto(
                updateDone);
        }
    }

    public record TenantStatusDto(
        bool StatusUpdated);
}

File: backoffice/src/server/queries/tenants/mutations/toggle-tenant-status.ts
import { fetchGraphQL } from "@/utils";

export async function toggleTenantStatusMutation(
  code: string,
  isEnabled: boolean
): Promise<boolean> {
  const query = `
    mutation ToggleTenantStatus($code: String!, $isEnabled: Boolean!) {
        toggleTenantStatus(command: {
            code: $code,
            isEnabled: $isEnabled
        }) {
            statusUpdated
        }
    }
`;

  const tenants = await fetchGraphQL<{ statusUpdated: boolean }>(query, {
    code,
    isEnabled,
  });

  return tenants.statusUpdated;
}
